File: python/tidegraph/full_rows.py
"""Full's affine rows batch independently of per-event nonlinear/Emit roots."""
import torch
from .isolated_linear import linear
from .full import FullResult


def evaluate(program, w, requests, slots, mode, zeta):
    from .ops import emit
    from . import lh_full
    if any(r.control.ndim != 0 for r in requests):
        raise ValueError("projection Emit requires scalar control")
    rows = [r.comparison.value for r in requests]
    if program.identity:
        fresh = [r.content.value for r in requests]
    elif w.full_kind in lh_full.PROFILES:
        fresh = [lh_full.fresh(w, row) for row in rows]
    else:
        if w.full_kind == "swiglu":
            gate, up = linear(rows, w.extra["ffn_gate"].t()), linear(rows, w.extra["ffn_up"].t())
            rows = linear([torch.nn.functional.silu(a)*b for a, b in zip(gate, up)], w.extra["ffn_down"].t())
        else:
            rows = [(v+w.bias).tanh() for v in linear(rows, w.weight.t())]
        fresh = [r.content.value+v for r, v in zip(requests, rows)]
    results = [FullResult(r.content.value if program.identity else emit(r.content.value, f, r.control, mode, zeta), {})
               for r, f in zip(requests, fresh)]
    for slot in range(slots):
        ids = [i for i, r in enumerate(requests) if program.present(slot, r.time)]
        if not ids:
            continue
        if program.kind == "broadcast":
            for i in ids:
                results[i].emitted[slot] = results[i].value
        else:
            weight = w.extra[f"emit_w_{slot}"]
            active = linear([fresh[i] for i in ids], weight.t())
            held = [requests[i].content.value for i in ids]
            if mode != "hard":
                held = linear(held, weight.t())
            for i, a, h in zip(ids, active, held):
                results[i].emitted[slot] = emit(h, a+w.extra[f"emit_b_{slot}"], requests[i].control, mode, zeta)
    return results

File: cpp/src/full_rows.cpp
#include "tide/full_rows.h"
#include "tide/isolated_linear.h"
#include "tide/lh_full.h"
#include "tide/operator_work.h"

namespace tide {
std::vector<Tensor> full_fresh_rows(const NodeWeights& w, const std::vector<FullInput>& inputs, bool identity) {
  std::vector<Tensor> rows;
  for (const auto& in : inputs) rows.push_back(identity || w.full_kind == "identity"
                                            ? in.content.value : in.comparison->value);
  if (identity || w.full_kind == "identity") return rows;
  if (is_lh_full(w.full_kind)) {
    for (auto& row : rows) row = lh_full_fresh(w, row);
    return rows;
  }
  const auto d = w.bias.numel();
  if (w.full_kind == "swiglu") {
    work::linear(work::FullCalls, rows.size(), d, 2*d);
    work::linear(work::FullCalls, rows.size(), d, 2*d);
    work::linear(work::FullCalls, rows.size(), 2*d, d);
    auto gate = isolated_linear(rows, w.extra.at("ffn_gate").t());
    auto up = isolated_linear(rows, w.extra.at("ffn_up").t());
    for (size_t i = 0; i < rows.size(); ++i) gate[i] = at::silu(gate[i])*up[i];
    rows = isolated_linear(gate, w.extra.at("ffn_down").t());
  } else {
    work::linear(work::FullCalls, rows.size(), d, d);
    rows = isolated_linear(rows, w.weight.t());
    for (auto& row : rows) row = at::tanh(row+w.bias);
  }
  for (size_t i = 0; i < rows.size(); ++i) rows[i] = inputs[i].content.value+rows[i];
  return rows;
}
}  // namespace tide
